Issue #61: Add Phaser in favor of a custom count and lock solution to Access and BufferLeaseImpl classes

eemhu opened the pull request:
#61 

Note: This PR is a draft

kortemik commented on src/main/java/com/teragrep/rlp_03/context/frame/rental/Rental.java:
```diff
@@ -0,0 +1,52 @@
+        phaser.arriveAndDeregister();
+    }
+
+    public boolean closed() {
```
is closed() extra?

kortemik commented on src/main/java/com/teragrep/rlp_03/context/buffer/BufferLeaseImpl.java:
```diff
@@ -51,55 +47,27 @@ public ByteBuffer buffer() {
     }
 
     @Override
     public boolean isRefCountZero() {
```
this must be changed so that it checks if the phaser isTerminated

eemhu replied:
fix 334d130d91728b1a8209c4d43affee26ed4c2baa

kortemik commented on src/main/java/com/teragrep/rlp_03/context/buffer/BufferLeaseImpl.java:
```diff
@@ -28,13 +29,8 @@ public long id() {
 
     @Override
     public long refs() {
```
this should be removed or at least marked as unreliable because phaser is a lock free structure so ref count should be only used for debug logging

kortemik commented on src/main/java/com/teragrep/rlp_03/context/frame/rental/Rental.java:
```diff
@@ -0,0 +1,52 @@
+public final class Rental implements AutoCloseable, Consumer<Lease>, Supplier<Lease> {
+
+    private final Phaser phaser;
+    private boolean closed;
```
this is extra, should return phaser.isTerminated() ?

eemhu replied:
return phaser.isTerminated: 334d130d91728b1a8209c4d43affee26ed4c2baa

kortemik commented on src/main/java/com/teragrep/rlp_03/context/frame/rental/Lease.java:
```diff
@@ -1,12 +1,12 @@
-package com.teragrep.rlp_03.context.frame.access;
+package com.teragrep.rlp_03.context.frame.rental;
 
 public class Lease implements AutoCloseable {
```
could this be refactored so that it uses a sub-phaser that is registered to the parent?

kortemik commented:
currently is broken, buffer is likely to be contaminated accross the threads:
```
[pool-9-thread-1] ERROR com.teragrep.rlp_03.context.RelpReadImpl - run() threw
java.lang.IllegalArgumentException: no match for EndOfTransfer character \n
        at com.teragrep.rlp_03.context.frame.function.EndOfTransferFunction.apply(EndOfTransferFunction.java:33)

```

i suspect the attemptRelease() is broken because clear() is not ensured by happens-before relationship of phaser.isTerminated()

kortemik commented:
there are test failures now

eemhu commented:
seem to pass tests locally, however "RentalTest contains IllegalStateException assert that does not seem to happen, commented out for now." in 38ec3d0

add sub phaser to Lease

kortemik commented:
Regarding BufferLeaseImpl, it might be appropriate to store cleared buffers
in the pool and take new Leases (with new phasers) for them when taking out
from the pool because phasers are terminated when returning.

pe 23. helmik. 2024 klo 10.52 eemhu ***@***.***> kirjoitti:

> seem to pass tests locally, however "RentalTest contains
> IllegalStateException assert that does not seem to happen, commented out
> for now." in 38ec3d0
> <https://github.com/teragrep/rlp_03/commit/38ec3d048a566c7d4ea3807a7106a504313b960b>
>
> add sub phaser to Lease
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/teragrep/rlp_03/pull/70#issuecomment-1960942329>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAGX2OTGWTTALPQULM5SL4LYVBKFRAVCNFSM6AAAAABDQYGTNSVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTSNRQHE2DEMZSHE>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

eemhu commented:
> Regarding BufferLeaseImpl, it might be appropriate to store cleared buffers in the pool and take new Leases (with new phasers) for them when taking out from the pool because phasers are terminated when returning. pe 23. helmik. 2024 klo 10.52 eemhu ***@***.***> kirjoitti:
> […](#)
> seem to pass tests locally, however "RentalTest contains IllegalStateException assert that does not seem to happen, commented out for now." in [38ec3d0](https://github.com/teragrep/rlp_03/commit/38ec3d048a566c7d4ea3807a7106a504313b960b) <[38ec3d0](https://github.com/teragrep/rlp_03/commit/38ec3d048a566c7d4ea3807a7106a504313b960b)> add sub phaser to Lease — Reply to this email directly, view it on GitHub <[#70 (comment)](https://github.com/teragrep/rlp_03/pull/70#issuecomment-1960942329)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AAGX2OTGWTTALPQULM5SL4LYVBKFRAVCNFSM6AAAAABDQYGTNSVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTSNRQHE2DEMZSHE> . You are receiving this because you commented.Message ID: ***@***.***>

Should `BufferLeasePool.take()` check the polled `BufferLease` for terminated `Phaser` and create a new `BufferLeaseImpl` in that case as well? Currently there is a null-check only.

kortemik commented:
I think it might be appropriate to use the decorator pattern for solving
this. One would have a decorator class with the phaser and do new
PhaserDecoratedBufferLeaseImpl(BufferLease bl) implements BufferLease.
Please create it so that only the terminated ones return the encapsulated
BufferLeaseImpl, otherwise a static stub is returned. This way I think the
code retains the responsibility assignment principle. I wonder if you think
this way too?

ma 26. helmik. 2024 klo 12.44 eemhu ***@***.***> kirjoitti:

> Regarding BufferLeaseImpl, it might be appropriate to store cleared
> buffers in the pool and take new Leases (with new phasers) for them when
> taking out from the pool because phasers are terminated when returning. pe
> 23. helmik. 2024 klo 10.52 eemhu *@*.
>
> *> kirjoitti: … <#m_-7089496626781693482_> seem to pass tests locally,
> however "RentalTest contains IllegalStateException assert that does not
> seem to happen, commented out for now." in 38ec3d0
> <https://github.com/teragrep/rlp_03/commit/38ec3d048a566c7d4ea3807a7106a504313b960b>
> <38ec3d0
> <https://github.com/teragrep/rlp_03/commit/38ec3d048a566c7d4ea3807a7106a504313b960b>>
> add sub phaser to Lease — Reply to this email directly, view it on GitHub
> <#70 (comment)
> <https://github.com/teragrep/rlp_03/pull/70#issuecomment-1960942329>>, or
> unsubscribe
> https://github.com/notifications/unsubscribe-auth/AAGX2OTGWTTALPQULM5SL4LYVBKFRAVCNFSM6AAAAABDQYGTNSVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTSNRQHE2DEMZSHE
> <https://github.com/notifications/unsubscribe-auth/AAGX2OTGWTTALPQULM5SL4LYVBKFRAVCNFSM6AAAAABDQYGTNSVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTSNRQHE2DEMZSHE>
> . You are receiving this because you commented.Message ID: @.*>
>
> Should BufferLeasePool.take() check the polled BufferLease for terminated
> Phaser and create a new BufferLeaseImpl in that case as well? Currently
> there is a null-check only.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/teragrep/rlp_03/pull/70#issuecomment-1963821583>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAGX2OQGEGMVRE4QFY63BODYVRRPPAVCNFSM6AAAAABDQYGTNSVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTSNRTHAZDCNJYGM>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

eemhu commented:
moved phaser to use decorator pattern `PhaserDecoratedBufferLease` which contains the phaser-related ops and `BufferLeaseImpl`contains the buffer itself.

kortemik commented:
Does it differ much from the Rental+Lease? If we'd generify Rental would it
fit?

ti 27. helmik. 2024 klo 15.47 eemhu ***@***.***> kirjoitti:

> moved phaser to use decorator pattern PhaserDecoratedBufferLease which
> contains the phaser-related ops and BufferLeaseImplcontains the buffer
> itself.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/teragrep/rlp_03/pull/70#issuecomment-1966588188>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAGX2OWL3FXV75F327DJ3ILYVXPWTAVCNFSM6AAAAABDQYGTNSVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTSNRWGU4DQMJYHA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

eemhu commented:
removed comparisons from phaser related classes, added synchronized keyword to bufferLeaseImpl.buffer() and removed lock. Add phaser decoration when polling from queue and remove when adding back to queue.

eemhu commented:
Perhaps the queue should take in type of PhaserDecoratedBufferLease so we can get rid of instanceof check?

kortemik commented:
Yes, this could be changed so that it accepts only decorated ones, as they
are the only type out there:

public void offer(BufferLease bufferLease) {

to 29. helmik. 2024 klo 10.28 eemhu ***@***.***> kirjoitti:

> Perhaps the queue should take in type of PhaserDecoratedBufferLease so we
> can get rid of instanceof check?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/teragrep/rlp_03/pull/70#issuecomment-1970643617>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAGX2OU4FI3IY3T76IRQKWLYV3TBXAVCNFSM6AAAAABDQYGTNSVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTSNZQGY2DGNRRG4>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

kortemik commented:
please separate the Access -> Rental to different pr, close this one and open new ones for these both.

kortemik commented:
running with the a1534ed following agains the ManualPerformanceTest:
```
/opt/Fail-Safe/rsyslog/rsyslog/bin/tcpflood -p1601 -y -Trelp-plain -m100000000 -c 8 -Y -M "<167>1 2003-03-01T01:00:00.000Z mymachine.example.com tcpflood - tag [tcpflood@32473 MSGNUM="12857358"] msgnum:"
```

gives
```
[pool-9-thread-4] ERROR com.teragrep.rlp_03.context.buffer.BufferLeasePool - dirty buffer in pool, terminating
```

while main branch does not.

@StrongestNumber9 please check the rlp_09 pr tests.

eemhu commented:
adding `synchronized` to BufferLeaseImpl.attemptRelease() method seems to fix it

eemhu commented:
would make sense considering the main branch version uses a lock there.

StrongestNumber9 commented:
> running with the [a1534ed](https://github.com/teragrep/rlp_03/commit/a1534edb12698618310dd33c30b5713cb55c19dc) following agains the ManualPerformanceTest:
> 
> ```
> /opt/Fail-Safe/rsyslog/rsyslog/bin/tcpflood -p1601 -y -Trelp-plain -m100000000 -c 8 -Y -M "<167>1 2003-03-01T01:00:00.000Z mymachine.example.com tcpflood - tag [tcpflood@32473 MSGNUM="12857358"] msgnum:"
> ```
> 
> gives
> 
> ```
> [pool-9-thread-4] ERROR com.teragrep.rlp_03.context.buffer.BufferLeasePool - dirty buffer in pool, terminating
> ```
> 
> while main branch does not.
> 
> @StrongestNumber9 please check the rlp_09 pr tests so that they will reproduce this.
> 
> @eemhu please check why this differs from main branch.
> 
> this utility is provided from rsyslog project via fail-safe-rsyslog-8.2102.2-1558.fc34.x86_64.rpm, one can obtain it from nexus.

Is this still relevant as 

> adding `synchronized` to BufferLeaseImpl.attemptRelease() method seems to fix it. Perhaps multiple threads try to access attemptRelease() at the same time and thus accessing the buffer as well.
